File: src/console.c
/*
 * ds-fork v6 - High-performance Container Runtime
 *
 * Copyright (C) 2026 ravindu644 <droidcasts@protonmail.com>
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

#include "ds-fork.h"

/* ---------------------------------------------------------------------------
 * Console Monitor Loop
 * ---------------------------------------------------------------------------*/

int console_monitor_loop(int master_fd, pid_t monitor_pid, struct config *cfg) {
  int epfd, sfd;
  sigset_t mask;
  struct signalfd_siginfo fdsi;
  struct epoll_event ev, events[10];
  char buf[4096];
  ssize_t n;
  int ret = 0;

  /* Pending write state for non-blocking PTY I/O.
   * When the container stops reading, the PTY buffer fills and write()
   * would block forever — deadlocking the entire event loop including
   * CTRL+ALT+Q.  We use non-blocking I/O + EPOLLOUT backpressure. */
  struct {
    int fd; /* target fd (master_fd or STDOUT_FILENO) */
    char data[4096];
    size_t len;
    size_t off;
  } pending = {.fd = -1};

  /* Setup signalfd for monitor signals */
  sigemptyset(&mask);
  sigaddset(&mask, SIGCHLD);
  sigaddset(&mask, SIGINT);
  sigaddset(&mask, SIGTERM);
  sigaddset(&mask, SIGWINCH);
  if (sigprocmask(SIG_BLOCK, &mask, NULL) < 0)
    return -1;

  sfd = signalfd(-1, &mask, SFD_NONBLOCK | SFD_CLOEXEC);
  if (sfd < 0)
    return -1;

  /* Setup epoll */
  epfd = epoll_create1(EPOLL_CLOEXEC);
  if (epfd < 0) {
    close(sfd);
    return -1;
  }

  /* 1. Watch user stdin */
  ev.events = EPOLLIN;
  ev.data.fd = STDIN_FILENO;
  if (epoll_ctl(epfd, EPOLL_CTL_ADD, STDIN_FILENO, &ev) < 0)
    log_warn("epoll_ctl(stdin) failed: %s", strerror(errno));

  /* 2. Watch PTY master (IN + HUP/ERR; OUT added only when pending data) */
  ev.events = EPOLLIN | EPOLLHUP | EPOLLERR;
  ev.data.fd = master_fd;
  if (epoll_ctl(epfd, EPOLL_CTL_ADD, master_fd, &ev) < 0)
    log_warn("epoll_ctl(master_fd) failed: %s", strerror(errno));

  /* 3. Watch signalfd */
  ev.events = EPOLLIN;
  ev.data.fd = sfd;
  if (epoll_ctl(epfd, EPOLL_CTL_ADD, sfd, &ev) < 0)
    log_warn("epoll_ctl(sig_fd) failed: %s", strerror(errno));

  /* Make PTY master non-blocking — the foundation of the backpressure fix */
  {
    int fl = fcntl(master_fd, F_GETFL);
    if (fl >= 0)
      fcntl(master_fd, F_SETFL, fl | O_NONBLOCK);
  }

  /* Set terminal to raw mode */
  struct termios oldtios;
  int is_tty = setup_tios(STDIN_FILENO, &oldtios);

  /* Initial window size sync */
  if (is_tty == 0) {
    struct winsize ws;
    if (ioctl(STDIN_FILENO, TIOCGWINSZ, &ws) == 0)
      ioctl(master_fd, TIOCSWINSZ, &ws);
  }

  int running = 1;
  while (running) {
    int nfds = epoll_wait(epfd, events, 10, -1);
    if (nfds < 0) {
      if (errno == EINTR)
        continue;
      ret = -1;
      break;
    }

    for (int i = 0; i < nfds; i++) {
      int fd = events[i].data.fd;

      if (fd == STDIN_FILENO) {
        /* User input -> Container master */
        n = read(STDIN_FILENO, buf, sizeof(buf));
        if (n > 0) {
          /* Check for CTRL+ALT+Q (\x1b\x11) escape sequence */
          if (n >= 2 && buf[0] == '\x1b' && buf[1] == '\x11') {
            static int exit_detected = 0;
            if (exit_detected == 0) {
              pid_t bg_pid = fork();
              if (bg_pid == 0) {
                setsid();
                log_silent = 1;
                stop_rootfs(cfg, 0);
                _exit(0);
              } else if (bg_pid > 0) {
                exit_detected = 1;
              }
            }
            continue;
          }

          /* Write to master_fd.  If the PTY buffer is full (container not
           * reading), buffer the data and register EPOLLOUT. */
          if (pending.fd < 0) {
            ssize_t w = write(master_fd, buf, (size_t)n);
            if (w >= 0 && (size_t)w < (size_t)n) {
              pending.fd = master_fd;
              pending.len = (size_t)n - (size_t)w;
              pending.off = 0;
              memcpy(pending.data, buf + w, pending.len);
              ev.events = EPOLLIN | EPOLLOUT | EPOLLHUP | EPOLLERR;
              ev.data.fd = master_fd;
              epoll_ctl(epfd, EPOLL_CTL_MOD, master_fd, &ev);
            } else if (w < 0 && errno == EAGAIN) {
              pending.fd = master_fd;
              pending.len = (size_t)n;
              pending.off = 0;
              memcpy(pending.data, buf, pending.len);
              ev.events = EPOLLIN | EPOLLOUT | EPOLLHUP | EPOLLERR;
              ev.data.fd = master_fd;
              epoll_ctl(epfd, EPOLL_CTL_MOD, master_fd, &ev);
            } else if (w < 0) {
              running = 0;
              break;
            }
          } else {
            /* Already have pending data — drop this input; container
             * is not consuming fast enough. */
          }
        }
      } else if (fd == master_fd) {
        if (events[i].events & (EPOLLHUP | EPOLLERR)) {
          running = 0;
          break;
        }

        /* Drain pending writes first (EPOLLOUT) */
        if (events[i].events & EPOLLOUT && pending.fd == master_fd) {
          ssize_t w = write(master_fd, pending.data + pending.off, pending.len);
          if (w > 0) {
            pending.off += (size_t)w;
            pending.len -= (size_t)w;
          }
          if (pending.len == 0 || (w < 0 && errno != EAGAIN)) {
            pending.fd = -1;
            ev.events = EPOLLIN | EPOLLHUP | EPOLLERR;
            ev.data.fd = master_fd;
            epoll_ctl(epfd, EPOLL_CTL_MOD, master_fd, &ev);
          }
        }

        /* Container output -> User stdout (EPOLLIN) */
        if (events[i].events & EPOLLIN) {
          n = read(master_fd, buf, sizeof(buf));
          if (n > 0) {
            ssize_t w = write(STDOUT_FILENO, buf, (size_t)n);
            (void)w; /* best-effort; partial is fine */
          } else {
            running = 0;
          }
        }
      } else if (fd == sfd) {
        /* Signal handling */
        n = read(sfd, &fdsi, sizeof(fdsi));
        if (n != sizeof(fdsi))
          continue;

        if (fdsi.ssi_signo == SIGCHLD) {
          int status;
          pid_t child = waitpid(monitor_pid, &status, WNOHANG);
          if (child == monitor_pid) {
            running = 0;
          }
        } else if (fdsi.ssi_signo == SIGWINCH) {
          struct winsize ws;
          if (ioctl(STDIN_FILENO, TIOCGWINSZ, &ws) == 0)
            ioctl(master_fd, TIOCSWINSZ, &ws);
        } else if (fdsi.ssi_signo == SIGINT || fdsi.ssi_signo == SIGTERM) {
          pid_t live_pid = find_container_init_pid(cfg->uuid);
          if (live_pid > 0)
            kill(live_pid, (int)fdsi.ssi_signo);
        }
      }
    }
  }

  /* Restore terminal settings */
  if (is_tty == 0) {
    tcsetattr(STDIN_FILENO, TCSAFLUSH, &oldtios);
  }

  close(epfd);
  close(sfd);
  return ret;
}

File: src/console.rs
//! 前台控制台 I/O 循环 —— 对应原 console.c。
//!
//! 基于 epoll 的 I/O 循环，转发 stdin → PTY master → stdout，
//! 支持非阻塞 PTY I/O 背压、CTRL+ALT+Q 退出检测、signalfd 信号处理。

use std::io;
use std::os::unix::io::RawFd;

use crate::container::stop_rootfs;
use crate::pid::find_container_init_pid;
use crate::terminal::setup_tios;
use crate::utils::set_log_silent;

// ══════════════════════════════════════════════════════════════════════════════
// 挂起写入状态
// ══════════════════════════════════════════════════════════════════════════════

struct PendingWrite {
    fd: RawFd,
    data: [u8; 4096],
    len: usize,
    off: usize,
}

impl PendingWrite {
    fn new() -> Self {
        Self { fd: -1, data: [0; 4096], len: 0, off: 0 }
    }
}

// ══════════════════════════════════════════════════════════════════════════════

pub fn console_monitor_loop(
    master_fd: RawFd,
    monitor_pid: libc::pid_t,
    cfg: &mut crate::types::Config,
) -> i32 {
    // ── signalfd 设置 ──
    let mut mask: libc::sigset_t = unsafe { std::mem::zeroed() };
    unsafe {
        libc::sigemptyset(&mut mask);
        libc::sigaddset(&mut mask, libc::SIGCHLD);
        libc::sigaddset(&mut mask, libc::SIGINT);
        libc::sigaddset(&mut mask, libc::SIGTERM);
        libc::sigaddset(&mut mask, libc::SIGWINCH);
    }
    if unsafe { libc::sigprocmask(libc::SIG_BLOCK, &mask, std::ptr::null_mut()) } < 0 {
        return -1;
    }

    let sfd = unsafe {
        libc::signalfd(-1, &mask, libc::SFD_NONBLOCK | libc::SFD_CLOEXEC)
    };
    if sfd < 0 { return -1; }

    // ── epoll 设置 ──
    let epfd = unsafe { libc::epoll_create1(libc::EPOLL_CLOEXEC) };
    if epfd < 0 { unsafe { libc::close(sfd) }; return -1; }

    let mut ev: libc::epoll_event = unsafe { std::mem::zeroed() };

    ev.events = libc::EPOLLIN as u32;
    ev.u64 = libc::STDIN_FILENO as u64;
    unsafe { libc::epoll_ctl(epfd, libc::EPOLL_CTL_ADD, libc::STDIN_FILENO, &mut ev); }

    ev.events = (libc::EPOLLIN | libc::EPOLLHUP | libc::EPOLLERR) as u32;
    ev.u64 = master_fd as u64;
    unsafe { libc::epoll_ctl(epfd, libc::EPOLL_CTL_ADD, master_fd, &mut ev); }

    ev.events = libc::EPOLLIN as u32;
    ev.u64 = sfd as u64;
    unsafe { libc::epoll_ctl(epfd, libc::EPOLL_CTL_ADD, sfd, &mut ev); }

    // PTY master 设为非阻塞
    let fl = unsafe { libc::fcntl(master_fd, libc::F_GETFL) };
    if fl >= 0 {
        unsafe { libc::fcntl(master_fd, libc::F_SETFL, fl | libc::O_NONBLOCK) };
    }

    // 终端设为 raw 模式
    let oldtios = setup_tios(libc::STDIN_FILENO);
    let is_tty = oldtios.is_ok();
    let oldtios = oldtios.unwrap_or_else(|_| unsafe { std::mem::zeroed() });

    // 初始窗口大小同步
    if is_tty {
        let mut ws: libc::winsize = unsafe { std::mem::zeroed() };
        if unsafe { libc::ioctl(libc::STDIN_FILENO, libc::TIOCGWINSZ, &mut ws) } == 0 {
            unsafe { libc::ioctl(master_fd, libc::TIOCSWINSZ, &ws) };
        }
    }

    let mut pending = PendingWrite::new();
    let mut running = true;
    let mut ret = 0;
    let mut exit_detected = false;
    let mut events: [libc::epoll_event; 10] = unsafe { std::mem::zeroed() };
    let mut buf = [0u8; 4096];

    while running {
        let nfds = unsafe { libc::epoll_wait(epfd, events.as_mut_ptr(), 10, -1) };
        if nfds < 0 {
            if io::Error::last_os_error().raw_os_error() == Some(libc::EINTR) {
                continue;
            }
            ret = -1;
            break;
        }

        for i in 0..nfds {
            let fd = events[i as usize].u64 as RawFd;

            if fd == libc::STDIN_FILENO {
                // 用户输入 → 容器 master
                let n = unsafe { libc::read(libc::STDIN_FILENO, buf.as_mut_ptr() as *mut libc::c_void, buf.len()) };
                if n > 0 {
                    // CTRL+ALT+Q (\x1b\x11) 退出序列
                    if n >= 2 && buf[0] == 0x1b && buf[1] == 0x11 {
                        if !exit_detected {
                            let bg_pid = unsafe { libc::fork() };
                            if bg_pid == 0 {
                                unsafe { libc::setsid() };
                                set_log_silent(true);
                                stop_rootfs(cfg, false);
                                unsafe { libc::_exit(0) };
                            } else if bg_pid > 0 {
                                exit_detected = true;
                            }
                        }
                        continue;
                    }

                    // 写入 master_fd，处理背压
                    let n_usize = n as usize;
                    if pending.fd < 0 {
                        let w = unsafe { libc::write(master_fd, buf.as_ptr() as *const libc::c_void, n_usize) };
                        if w >= 0 && (w as usize) < n_usize {
                            pending.fd = master_fd;
                            pending.len = n_usize - w as usize;
                            pending.off = 0;
                            pending.data[..pending.len].copy_from_slice(&buf[w as usize..n_usize]);
                            add_epollout(epfd, master_fd);
                        } else if w < 0 {
                            let err = io::Error::last_os_error();
                            if err.raw_os_error() == Some(libc::EAGAIN) {
                                pending.fd = master_fd;
                                pending.len = n_usize;
                                pending.off = 0;
                                pending.data[..pending.len].copy_from_slice(&buf[..n_usize]);
                                add_epollout(epfd, master_fd);
                            } else {
                                running = false;
                                break;
                            }
                        }
                    }
                }
            } else if fd == master_fd {
                if events[i as usize].events & (libc::EPOLLHUP as u32 | libc::EPOLLERR as u32) != 0 {
                    running = false;
                    break;
                }

                // 排空挂起的写入（EPOLLOUT）
                if events[i as usize].events & libc::EPOLLOUT as u32 != 0 && pending.fd == master_fd {
                    let w = unsafe {
                        libc::write(master_fd,
                            pending.data[pending.off..].as_ptr() as *const libc::c_void,
                            pending.len)
                    };
                    if w > 0 {
                        pending.off += w as usize;
                        pending.len -= w as usize;
                    }
                    if pending.len == 0 || (w < 0 && io::Error::last_os_error().raw_os_error() != Some(libc::EAGAIN)) {
                        pending.fd = -1;
                        del_epollout(epfd, master_fd);
                    }
                }

                // 容器输出 → 用户 stdout（EPOLLIN）
                if events[i as usize].events & libc::EPOLLIN as u32 != 0 {
                    let n = unsafe { libc::read(master_fd, buf.as_mut_ptr() as *mut libc::c_void, buf.len()) };
                    if n > 0 {
                        unsafe { libc::write(libc::STDOUT_FILENO, buf.as_ptr() as *const libc::c_void, n as usize) };
                    } else {
                        running = false;
                    }
                }
            } else if fd == sfd {
                // 信号处理
                let mut fdsi: libc::signalfd_siginfo = unsafe { std::mem::zeroed() };
                let n = unsafe {
                    libc::read(sfd, &mut fdsi as *mut libc::signalfd_siginfo as *mut libc::c_void, std::mem::size_of::<libc::signalfd_siginfo>())
                };
                if n != std::mem::size_of::<libc::signalfd_siginfo>() as isize { continue; }

                if fdsi.ssi_signo == libc::SIGCHLD as u32 {
                    let mut status = 0;
                    let child = unsafe { libc::waitpid(monitor_pid, &mut status, libc::WNOHANG) };
                    if child == monitor_pid { running = false; }
                } else if fdsi.ssi_signo == libc::SIGWINCH as u32 {
                    let mut ws: libc::winsize = unsafe { std::mem::zeroed() };
                    if unsafe { libc::ioctl(libc::STDIN_FILENO, libc::TIOCGWINSZ, &mut ws) } == 0 {
                        unsafe { libc::ioctl(master_fd, libc::TIOCSWINSZ, &ws) };
                    }
                } else if (fdsi.ssi_signo == libc::SIGINT as u32 || fdsi.ssi_signo == libc::SIGTERM as u32)
                    && !cfg.uuid.is_empty() {
                        let live_pid = find_container_init_pid(&cfg.uuid);
                        if live_pid > 0 {
                            unsafe { libc::kill(live_pid, fdsi.ssi_signo as i32) };
                        }
                    }
            }
        }
    }

    // 恢复终端设置
    if is_tty {
        unsafe { libc::tcsetattr(libc::STDIN_FILENO, libc::TCSAFLUSH, &oldtios) };
    }

    unsafe {
        libc::close(epfd);
        libc::close(sfd);
    }
    ret
}

fn add_epollout(epfd: RawFd, fd: RawFd) {
    let mut ev: libc::epoll_event = unsafe { std::mem::zeroed() };
    ev.events = (libc::EPOLLIN | libc::EPOLLOUT | libc::EPOLLHUP | libc::EPOLLERR) as u32;
    ev.u64 = fd as u64;
    unsafe { libc::epoll_ctl(epfd, libc::EPOLL_CTL_MOD, fd, &mut ev) };
}

fn del_epollout(epfd: RawFd, fd: RawFd) {
    let mut ev: libc::epoll_event = unsafe { std::mem::zeroed() };
    ev.events = (libc::EPOLLIN | libc::EPOLLHUP | libc::EPOLLERR) as u32;
    ev.u64 = fd as u64;
    unsafe { libc::epoll_ctl(epfd, libc::EPOLL_CTL_MOD, fd, &mut ev) };
}
